kvdb/postgres/readwrite_tx.go

```diff
@@ -0,0 +1,296 @@
+package postgres
+
+import (
+	"bytes"
+	"database/sql"
+	"sort"
+	"strings"
+
+	"github.com/btcsuite/btcwallet/walletdb"
+)
+
+// readWriteTx holds a reference to an open postgres transaction.
+type readWriteTx struct {
+	readOnly bool
+	db       *db
+	tx       *sql.Tx
+
+	// onCommit gets called upon commit.
+	onCommit func()
+
+	// active is true if the transaction hasn't been committed yet.
+	active bool
+}
+
+// newReadWriteTx creates an rw transaction using a connection from the
+// specified pool.
+func newReadWriteTx(db *db, readOnly bool) (*readWriteTx, error) {
+	// Obtain the global lock instance. An alternative here is to obtain a
+	// database lock from Postgres. Unfortunately there is no database-level
+	// lock in Postgres, meaning that each table would need to be locked
+	// individually. Perhaps an advisory lock could perform this function
+	// too.
+	if readOnly {
+		db.lock.RLock()
+	} else {
+		db.lock.Lock()
+	}
+
+	tx, err := db.db.Begin()
+	if err != nil {
+		return nil, err
```

Comment:
Shouldn't we unlock here to avoid deadlock?

Reply:
good point, fixed.